File: php/src/EBNF/ast/AbstractNode.php
<?php

/**
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @license http://www.gnu.org/licenses/ GNU General Public License
 * @author  Sven Strittmatter <ich@weltraumschaf.de>
 * @package ast
 */

namespace de\weltraumschaf\ebnf\ast;

/**
 * @see Node
 */
require_once __DIR__ . DIRECTORY_SEPARATOR . 'Node.php';

/**
 * Abstract representation of AST nodes which are not the root
 * node {@link Syntax} but any other kind of {@link Node}.
 *
 * @package ast
 * @version @@version@@
 */
abstract class AbstractNode implements Node {

    /**
     * The direct ancestor int the AST tree.
     *
     * @var Node
     */
    private $parent;

    /**
     * Initializes the node with it's parent. This is imutable.
     *
     * May the node itself change, it is not possible to set the
     * reference to the parent node.
     *
     * @param Node $parent Ancestor node.
     */
    public function __construct(Node $parent) {
        $this->parent = $parent;
    }

    /**
     * Returns the parent node.
     *
     * @return Node
     */
    public function getParent() {
        return $this->parent;
    }
}

File: java/src/main/java/de/weltraumschaf/ebnf/ast/AbstractNode.java
package de.weltraumschaf.ebnf.ast;

import com.google.common.collect.Maps;
import de.weltraumschaf.ebnf.visitor.Visitor;
import java.util.Map;

/**
 * Abstract representation of AST nodes which are not the {@link nodes.Syntax "root node"}
 * but any other kind of {@link Node}.
 *
 * @author Sven Strittmatter <weltraumschaf@googlemail.com>
 */
public abstract class AbstractNode implements Node {

    /**
     * The direct ancestor in the AST tree.
     */
    private final Node parent;
    private final NodeType type;
    private final Map<String, String> attributes;

    /**
     * Initializes the node with it's parent. This is immutable.
     *
     * May the node itself change, it is not possible to set the
     * reference to the parent node.
     *
     * @param Node Ancestor node.
     */
    protected AbstractNode(final Node parent, final NodeType type) {
        this.parent     = parent;
        this.type       = type;
        this.attributes = Maps.newHashMap();
    }

    @Override
    public boolean hasParent() {
        return parent != null && parent.getType() != NodeType.NULL;
    }

    /**
     * Returns the parent node.
     *
     * @return
     */
    @Override
    public Node getParent() {
        return parent;
    }

    @Override
    public NodeType getType() {
        return type;
    }

    @Override
    public String getNodeName() {
        return getType().toString();
    }

    /**
     * Defines method to accept {@link Visitor visitors}.
     *
     * Implements <a href="http://en.wikipedia.org/wiki/Visitor_pattern">Visitor Pattern</a>.
     *
     * @param visitor Object which visits the node.
     */
    @Override
    public void accept(final Visitor visitor) {
        visitor.beforeVisit(this);
        visitor.visit(this);
        visitor.afterVisit(this);
    }

    @Override
    public boolean hasAttributes() {
        return !attributes.isEmpty();
    }

    @Override
    public boolean hasAttribute(final String name) {
        return attributes.containsKey(name);
    }

    @Override
    public String getAttribute(final String name) {
        if (!hasAttribute(name)) {
            throw new IllegalArgumentException(String.format("Does not have attribute with name '%s'!", name));
        }

        return attributes.get(name);
    }

    @Override
    public void setAttribute(final String name, final String value) {
        attributes.put(name, value);
    }

}
